internal/events/block_confirmations.go

```diff
@@ -0,0 +1,520 @@
+	defaultEventQueueLength = 100
+	defaultPollingInterval  = 1 * time.Second
+)
+
+// bcmConfExternal is the YAML/JSON parsable external configuration, with pointers to allow defaults to be applied, and units
+type bcmConfExternal struct {
+	Enabled                 bool `json:"enabled,omitempty"`
+	RequiredConfirmations   *int `json:"requiredConfirmations,omitempty"`
+	BlockCacheSize          *int `json:"blockCacheSize,omitempty"`
+	BlockPollingIntervalSec *int `json:"pollingIntervalSec,omitempty"`
+	EventQueueLength        *int `json:"eventQueueLength,omitempty"`
+	IncludeInPayload        bool `json:"includeInPayload,omitempty"`
+}
+
+// bcmConfInternal is the parsed config ready to use
+type bcmConfInternal struct {
+	requiredConfirmations int
+	blockCacheSize        int
+	pollingInterval       time.Duration
+	eventQueueLength      int
+	includeInPayload      bool
+}
+
+type pendingEvent struct {
+	key           string
+	added         time.Time
+	confirmations []*blockInfo
+	event         *eventData
+	eventStream   *eventStream
+}
+
+type pendingEvents []*pendingEvent
+
+func (pe pendingEvents) Len() int      { return len(pe) }
+func (pe pendingEvents) Swap(i, j int) { pe[i], pe[j] = pe[j], pe[i] }
+func (pe pendingEvents) Less(i, j int) bool {
+	return pe[i].event.blockNumber < pe[j].event.blockNumber ||
+		(pe[i].event.blockNumber == pe[j].event.blockNumber && (pe[i].event.transactionIndex < pe[j].event.transactionIndex ||
+			(pe[i].event.transactionIndex == pe[j].event.transactionIndex && pe[i].event.logIndex < pe[j].event.logIndex)))
+}
+
+type bcmEventType int
+
+const (
+	bcmNewLog bcmEventType = iota
+	bcmRemovedLog
+	bcmStopStream
+)
+
+type bcmNotification struct {
+	nType       bcmEventType
+	event       *eventData
+	eventStream *eventStream
+	complete    chan struct{} // for bcmStopStream only
+}
+
+// blockInfo is the information we cache for a block
+type blockInfo struct {
+	Number     ethbinding.HexUint `json:"number"`
+	Hash       ethbinding.Hash    `json:"hash"`
+	ParentHash ethbinding.Hash    `json:"parentHash"`
+	Timestamp  ethbinding.HexUint `json:"timestamp"`
+}
+
+// MarshalJSON for more consistent JSON output in confirmations array of event
+func (bi *blockInfo) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]interface{}{
+		"number":     strconv.FormatUint(uint64(bi.Number), 10),
+		"hash":       bi.Hash,
+		"parentHash": bi.ParentHash,
+		"timestamp":  strconv.FormatUint(uint64(bi.Timestamp), 10),
+	})
+}
+
+func parseBCMConfig(conf *bcmConfExternal) *bcmConfInternal {
+	intConf := &bcmConfInternal{
+		requiredConfirmations: defaultConfirmations,
+		blockCacheSize:        defaultBlockCacheSize,
+		pollingInterval:       defaultPollingInterval,
+		eventQueueLength:      defaultEventQueueLength,
+	}
+	if conf.RequiredConfirmations != nil {
+		intConf.requiredConfirmations = *conf.RequiredConfirmations
+	}
+	if conf.BlockCacheSize != nil {
+		intConf.blockCacheSize = *conf.BlockCacheSize
```

Comment:
should we enforce a safe upper bound (no idea what that should be though!)?

Reply:
I guess we haven't elsewhere on these. I propose no is ok for now